contrib/platform/test/com/sun/jna/platform/win32/Kernel32Test.java

```diff
@@ -1910,4 +1912,53 @@ public void run() {
 
         assertNull(mutexHandle);
     }
+
+    public void testApplicationRestart() {
+        try {
+            HRESULT insufficientBuffer = W32Errors.HRESULT_FROM_WIN32(WinError.ERROR_INSUFFICIENT_BUFFER);
+            HRESULT result;
+            String dummyCommandline = "/restart -f .\\filename.ext";
+            char[] dummyCommandlineArray = Native.toCharArray(dummyCommandline);
+            int dummyFlags = 2;
+            result = Kernel32.INSTANCE.RegisterApplicationRestart(dummyCommandlineArray, dummyFlags);
+            assertTrue(COMUtils.SUCCEEDED(result));
+            
+            char[] queriedCommandlineArray = null;
+            IntByReference queriedSize = new IntByReference();
+            IntByReference queriedFlags = new IntByReference();
+
+            // Query without target buffer to determine required buffer size
+            result = Kernel32.INSTANCE.GetApplicationRestartSettings(Kernel32.INSTANCE.GetCurrentProcess(), queriedCommandlineArray, queriedSize, queriedFlags);
+
+            assertTrue(COMUtils.SUCCEEDED(result));
+            assertEquals(dummyCommandline.length() + 1, queriedSize.getValue());
+
+            // Check error reporting, use insufficient buffer size
+            queriedCommandlineArray = new char[1];
+            queriedSize.setValue(queriedCommandlineArray.length);
+            queriedFlags.setValue(-1);
+            result = Kernel32.INSTANCE.GetApplicationRestartSettings(Kernel32.INSTANCE.GetCurrentProcess(), queriedCommandlineArray, queriedSize, queriedFlags);
+
+            assertTrue(COMUtils.FAILED(result));
+            assertEquals(insufficientBuffer, result);
+            assertEquals(dummyCommandline.length() + 1, queriedSize.getValue());
+
+            // Now query with the right buffer size
+            queriedCommandlineArray = new char[queriedSize.getValue()];
+            queriedSize.setValue(queriedCommandlineArray.length);
+            queriedFlags.setValue(-1);
+            result = Kernel32.INSTANCE.GetApplicationRestartSettings(Kernel32.INSTANCE.GetCurrentProcess(), queriedCommandlineArray, queriedSize, queriedFlags);
+
+            assertTrue(COMUtils.SUCCEEDED(result));
+            assertEquals(dummyCommandline.length() + 1, queriedSize.getValue());
+            assertEquals(dummyFlags, queriedFlags.getValue());
+            assertEquals(dummyCommandline, Native.toString(queriedCommandlineArray));
+            
+            result = Kernel32.INSTANCE.UnregisterApplicationRestart();
+            assertTrue(COMUtils.SUCCEEDED(result));
+        } finally {
+            // Last resort if test succeeds partially
+            Kernel32.INSTANCE.UnregisterApplicationRestart();
```

Comment:
The `try` block should be around after a successful `RegisterApplicationRestart`. Then you can combine the `UnregisterApplicationRestart` call and  `assertTrue(COMUtils.SUCCEEDED(result));` here so that any errors bubble up.

Reply:
This is not the intention here. `UnregisterApplicationRestart` is only run in the finally block, so that cleanup happens in any case. The real testing happens inside the the try-block. The real test for `UnregisterApplicationRestart` is in 1957.

Reply:
I am possibly nitpicking, so feel free to leave as is. However this is what I am thinking the issues are and generally avoid the practice of doing what you're doing:

- if `RegisterApplicationRestart` succeeds in 1923, finally calls `UnregisterApplicationRestart` a second time, which fails silently, which seems like bad practice.
- if `RegisterApplicationRestart` fails in 1923, finally calls `UnregisterApplicationRestart` a first time, which fails silently, which seems like bad practice.
- if `RegisterApplicationRestart` succeeds in 1923, but something else fails, finally calls `UnregisterApplicationRestart` a first time, which may fail silently, which will leave a registered application restart without anyone noticing, which might potentially cause another subsequent test to fail.

Happy holidays!

Reply:
I'll leave as is. the reasoning is, that this is a unittest not more, not less. The only intention of the finally block is to cleanup as much as possible. It is a last resort - if it works, it will clean up, if if fails, we loose nothing. The functions themselves are tested inside the try-block.